[AMEND] Updated the redirect URL Repository to search for entries wit…

NikRimington opened the pull request:
This fix is to address issue #9636 . If a search based on culture fails to return any redirects it will search for any that have no culture defined against them.

I feel that an additional unit test is probably required, but I wasn't sure how to set up the data for the test so have left this off.

I did consider falling back to the original method that doesn't look at culture but I felt that using "no culture" would be a better approach.

Shazwazza commented:
@NikRimington there's an existing `RedirectUrlRepositoryTests`, would you be able to add tests there?

NikRimington commented:
Hi @Shazwazza I've added in a new test for the fallback behaviour which appears to be passing. I also cleaned up the tests a little bit so variables had more sensible names as I think it added to my confusion on how to create the new test. Hope that's okay :-)

Shazwazza commented on src/Umbraco.Core/Persistence/Repositories/Implement/RedirectUrlRepository.cs:
```diff
@@ -218,7 +218,21 @@ public IRedirectUrl GetMostRecentUrl(string url, string culture)
+
+        private RedirectUrlDto GetMostRecentUrlWithNoCulture(string url)
+        {
+            var urlHash = url.GenerateHash<SHA1>();
```
This is regenerating a hash that was already generated. The url hasn't changed so no need to allocate and add overhead hashing again.

NikRimington replied:
This function has now been removed as isn't needed any more :-)

Shazwazza commented on src/Umbraco.Core/Persistence/Repositories/Implement/RedirectUrlRepository.cs:
```diff
@@ -218,7 +218,21 @@ public IRedirectUrl GetMostRecentUrl(string url, string culture)
+        {
+            var urlHash = url.GenerateHash<SHA1>();
+            var sql = GetBaseQuery(false)
+                .Where<RedirectUrlDto>(x => x.Url == url && x.UrlHash == urlHash && x.Culture == string.Empty)
```
Is there any reason why we can't 'just' have the single query above check `x.Culture == culture.ToLower() || x.Culture == string.Empty)` ? Would be better to have less queries if possible.

NikRimington replied:
I've combined it into a single DB query, but this means there is filtering on the Linq expression (FirstOrDefault) now. I hope this is okay?

Shazwazza reviewed (changes requested):
Have added a couple note and questions

nul800sebastiaan commented:
Thanks @bjarnef - I've cherry-picked for 8.10 and 8.11 👍

nul800sebastiaan commented:
Duh.. Wrong PR ☝️ 

I've tested this and it now works. I've cherry-picked for 8.10 and 8.11 👍
Thanks for the excellent help @NikRimington! 🎉
